src/watch_state.rs

```diff
@@ -0,0 +1,231 @@
+use std::sync::{Arc, Mutex};
+
+use watchexec::action::Outcome;
+
+/// Tracks the current state of spin watch. Handles transitioning states and converting effects
+/// into outcomes based on the current state.
+#[derive(Debug, Clone)]
+pub struct StateMachine {
+    state: Arc<Mutex<State>>,
+    clear: bool,
+    skip_build: bool,
+}
+
+impl StateMachine {
+    pub fn new(skip_build: bool, clear: bool) -> Self {
+        Self {
+            state: Arc::new(Mutex::new(match skip_build {
+                false => State::Building,
+                true => State::Running,
+            })),
+            clear,
+            skip_build,
+        }
+    }
+
+    /// Get the current state of the state machine.
+    pub fn get_state(&self) -> State {
+        *self.state.lock().unwrap()
+    }
+
+    /// Based on the given effect return the correct outcome. Transition the internal state of the
+    /// machine if necessary.
+    pub fn handle(&self, effect: Effect) -> Outcome {
+        let mut state = self.state.lock().unwrap();
+        tracing::debug!("handling effect {:?} in current state {:?}", effect, *state);
+        let outcome = match (effect, *state) {
```

Comment:
I really like the clarity (and testability) of the state machine approach.  Very nice.

Reply:
Thanks. It was really fun to design and code.